### PR TITLE
Enable local-only mode for scraping options

### PR DESCRIPTION
**Pull request by lat0z**
This change adds a flag to limit the scraping process to the local node only, this might be useful in some scenarios where it is preferred to deploy the exporter to all or few nodes within a cluster rather than a single one. 
We have tested this option within 10+ different clusters, and so far we have reduced the n^2 amount of metrics within each node for the couchdb_httpd* and couchdb_server* metrics

### CODE REVIEW

**Comment by gesellix** [comment]
Hey @lat0z, your pull request is not forgotten, I'll come back to it in a few days!

File: README.md [hunk 1]
```diff
@@ -92,6 +92,10 @@ credentials, e.g. like this:
 
     couchdb-prometheus-exporter --couchdb.uri=http://couchdb:5984 --couchdb.username=root --couchdb.password=a-secret
 
+It is also possible to fetch the stats from a single node, using the scrape.local option, e.g. Like this:
+
+    couchdb-prometheus-exporter --couchdb.uri=http://couchdb:5984 --couchdb.username=root --couchdb.password=a-secret --scrape.localnode=true
```
Comment (gesellix) [inline README.md:97]:
```suggestion
It is also possible to fetch the stats from a single node, using the scrape.localonly option, e.g. Like this:

    couchdb-prometheus-exporter --couchdb.uri=http://couchdb:5984 --couchdb.username=root --couchdb.password=a-secret --scrape.localonly=true
```

File: couchdb-exporter.go [hunk 2]
```diff
@@ -36,6 +36,7 @@ type exporterConfigType struct {
 	couchdbPassword            string
 	couchdbInsecure            bool
 	scrapeInterval             time.Duration
+	scrapeLocalNode            bool
```
Comment (gesellix) [inline couchdb-exporter.go:39]:
```suggestion
	scrapeLocalOnly            bool
```

File: couchdb-exporter.go [hunk 3]
```diff
@@ -126,6 +127,14 @@ func init() {
 			Value:       0 * time.Second,
 			Destination: &exporterConfig.scrapeInterval,
 		}),
+		altsrc.NewBoolFlag(&cli.BoolFlag{
+			Name:        "scrape.localnode",
```
Comment (gesellix) [inline couchdb-exporter.go:131]:
```suggestion
			Name:        "scrape.localonly",
```

File: couchdb-exporter.go [hunk 4]
```diff
@@ -126,6 +127,14 @@ func init() {
...
+			Usage:       fmt.Sprintf("Whether collect metrics from the whole cluster or local instance"),
+			EnvVars:     []string{"SCRAPE_MODE"},
```
Comment (gesellix) [inline couchdb-exporter.go:133]:
```suggestion
			EnvVars:     []string{"SCRAPE_LOCAL_ONLY"},
```

File: couchdb-exporter.go [hunk 5]
```diff
@@ -126,6 +127,14 @@ func init() {
...
+			Hidden:      false,
+			Value:       false,
+			Destination: &exporterConfig.scrapeLocalNode,
```
Comment (gesellix) [inline couchdb-exporter.go:136]:
```suggestion
			Destination: &exporterConfig.scrapeLocalOnly,
```

File: couchdb-exporter.go [hunk 6]
```diff
@@ -179,6 +188,7 @@ func main() {
 
 		exporter := lib.NewExporter(
 			exporterConfig.couchdbURI,
+			exporterConfig.scrapeLocalNode,
```
Comment (gesellix) [inline couchdb-exporter.go:191]:
```suggestion
			exporterConfig.scrapeLocalOnly,
```

File: couchdb-exporter_test.go [hunk 7]
```diff
@@ -125,8 +125,9 @@ func performCouchdbStatsTest(t *testing.T, scrapeInterval time.Duration, couchdb
 	basicAuth := lib.BasicAuth{Username: "username", Password: "password"}
 	handler := http.HandlerFunc(BasicAuthHandler(basicAuth, couchdbResponse(t, couchdbVersion)))
 	server := httptest.NewServer(handler)
+	localMode:= false
```
Comment (gesellix) [inline couchdb-exporter_test.go:128]:
```suggestion
	localOnly := false
```

File: couchdb-exporter_test.go [hunk 8]
```diff
@@ -125,8 +125,9 @@ func performCouchdbStatsTest(t *testing.T, scrapeInterval time.Duration, couchdb
...
 
-	e := lib.NewExporter(server.URL, basicAuth, lib.CollectorConfig{
+	e := lib.NewExporter(server.URL, localMode, basicAuth, lib.CollectorConfig{
```
Comment (gesellix) [inline couchdb-exporter_test.go:130]:
```suggestion
	e := lib.NewExporter(server.URL, localOnly, basicAuth, lib.CollectorConfig{
```

File: couchdb-exporter_test.go [hunk 9]
```diff
@@ -208,8 +209,9 @@ func TestCouchdbStatsV1Integration(t *testing.T) {
 	}
 	dbUrl := fmt.Sprintf("http://%s", dbAddress)
 	basicAuth := lib.BasicAuth{Username: "root", Password: "a-secret"}
+	localMode := false
```
Comment (gesellix) [inline couchdb-exporter_test.go:212]:
```suggestion
	localOnly := false
```

File: couchdb-exporter_test.go [hunk 10]
```diff
@@ -208,8 +209,9 @@ func TestCouchdbStatsV1Integration(t *testing.T) {
...
 
-	client := lib.NewCouchdbClient(dbUrl, basicAuth, true)
+	client := lib.NewCouchdbClient(dbUrl, localMode, basicAuth, true)
```
Comment (gesellix) [inline couchdb-exporter_test.go:214]:
```suggestion
	client := lib.NewCouchdbClient(dbUrl, localOnly, basicAuth, true)
```

File: couchdb-exporter_test.go [hunk 11]
```diff
@@ -220,7 +222,7 @@ func TestCouchdbStatsV1Integration(t *testing.T) {
 
 	scrapeInterval, _ := time.ParseDuration("0s")
 	t.Run("node_up", func(t *testing.T) {
-		e := lib.NewExporter(dbUrl, basicAuth, lib.CollectorConfig{
+		e := lib.NewExporter(dbUrl, localMode, basicAuth, lib.CollectorConfig{
```
Comment (gesellix) [inline couchdb-exporter_test.go:225]:
```suggestion
		e := lib.NewExporter(dbUrl, localOnly, basicAuth, lib.CollectorConfig{
```

File: couchdb-exporter_test.go [hunk 12]
```diff
@@ -245,7 +247,7 @@ func TestCouchdbStatsV1Integration(t *testing.T) {
 	})
 
 	t.Run("_all_dbs", func(t *testing.T) {
-		e := lib.NewExporter(dbUrl, basicAuth, lib.CollectorConfig{
+		e := lib.NewExporter(dbUrl, localMode, basicAuth, lib.CollectorConfig{
```
Comment (gesellix) [inline couchdb-exporter_test.go:250]:
```suggestion
		e := lib.NewExporter(dbUrl, localOnly, basicAuth, lib.CollectorConfig{
```

File: couchdb-exporter_test.go [hunk 13]
```diff
@@ -279,13 +281,13 @@ func TestCouchdbStatsV1Integration(t *testing.T) {
 	}
 }
 
-func awaitMembership(t *testing.T, basicAuth lib.BasicAuth) func(address string) (bool, error) {
+func awaitMembership(t *testing.T,localMode bool, basicAuth lib.BasicAuth) func(address string) (bool, error) {
```
Comment (gesellix) [inline couchdb-exporter_test.go:284]:
```suggestion
func awaitMembership(t *testing.T, localOnly bool, basicAuth lib.BasicAuth) func(address string) (bool, error) {
```

File: couchdb-exporter_test.go [hunk 14]
```diff
@@ -279,13 +281,13 @@ func TestCouchdbStatsV1Integration(t *testing.T) {
...
 	time.Sleep(5 * time.Second)
 
 	return func(address string) (bool, error) {
 		dbUrl := fmt.Sprintf("http://%s", address)
-		c := lib.NewCouchdbClient(dbUrl, basicAuth, true)
-		nodeNames, err := c.GetNodeNames()
+		c := lib.NewCouchdbClient(dbUrl, localMode, basicAuth, true)
+		nodeNames, err := c.GetNodeNames(localMode)
```
Comment (gesellix) [inline couchdb-exporter_test.go:290]:
```suggestion
		c := lib.NewCouchdbClient(dbUrl, localOnly, basicAuth, true)
		nodeNames, err := c.GetNodeNames(localOnly)
```

File: couchdb-exporter_test.go [hunk 15]
```diff
@@ -303,6 +305,28 @@ func awaitMembership(t *testing.T, basicAuth lib.BasicAuth) func(address string)
 	}
 }
 
+func awaitLocalNode(t *testing.T, localMode bool, basicAuth lib.BasicAuth) func(address string) (bool, error) {
```
Comment (gesellix) [inline couchdb-exporter_test.go:308]:
```suggestion
func awaitLocalNode(t *testing.T, localOnly bool, basicAuth lib.BasicAuth) func(address string) (bool, error) {
```

File: couchdb-exporter_test.go [hunk 16]
```diff
@@ -303,6 +305,28 @@ func awaitMembership(t *testing.T, basicAuth lib.BasicAuth) func(address string)
...
+	time.Sleep(5 * time.Second)
+
+	return func(address string) (bool, error) {
+		dbUrl := fmt.Sprintf("http://%s", address)
+		c := lib.NewCouchdbClient(dbUrl, localMode, basicAuth, true)
```
Comment (gesellix) [inline couchdb-exporter_test.go:313]:
```suggestion
		c := lib.NewCouchdbClient(dbUrl, localOnly, basicAuth, true)
```

File: couchdb-exporter_test.go [hunk 17]
```diff
@@ -303,6 +305,28 @@ func awaitMembership(t *testing.T, basicAuth lib.BasicAuth) func(address string)
...
+		nodeNames, err := c.GetNodeNames(localMode)
```
Comment (gesellix) [inline couchdb-exporter_test.go:314]:
```suggestion
		nodeNames, err := c.GetNodeNames(localOnly)
```

File: couchdb-exporter_test.go [hunk 18]
```diff
@@ -317,13 +341,19 @@ func TestCouchdbStatsV2Integration(t *testing.T) {
 
 	dbUrl := fmt.Sprintf("http://%s", dbAddress)
 	basicAuth := lib.BasicAuth{Username: "root", Password: "a-secret"}
+	localMode := false
```
Comment (gesellix) [inline couchdb-exporter_test.go:344]:
```suggestion
	localOnly := false
```

File: couchdb-exporter_test.go [hunk 19]
```diff
@@ -317,13 +341,19 @@ func TestCouchdbStatsV2Integration(t *testing.T) {
...
+
+	err = cluster_config.AwaitNodes([]string{dbAddress}, clusterSetupDelay, clusterSetupTimeout, awaitLocalNode(t, true, basicAuth))
+	if err != nil {
+		t.Error(err)
+	}
 
-	err = cluster_config.AwaitNodes([]string{dbAddress}, clusterSetupDelay, clusterSetupTimeout, awaitMembership(t, basicAuth))
+	err = cluster_config.AwaitNodes([]string{dbAddress}, clusterSetupDelay, clusterSetupTimeout, awaitMembership(t, localMode, basicAuth))
```
Comment (gesellix) [inline couchdb-exporter_test.go:351]:
```suggestion
	err = cluster_config.AwaitNodes([]string{dbAddress}, clusterSetupDelay, clusterSetupTimeout, awaitMembership(t, localOnly, basicAuth))
```

File: couchdb-exporter_test.go [hunk 20]
```diff
@@ -317,13 +341,19 @@ func TestCouchdbStatsV2Integration(t *testing.T) {
...
 	if err != nil {
 		t.Error(err)
 	}
 
-	client := lib.NewCouchdbClient(dbUrl, basicAuth, true)
+	client := lib.NewCouchdbClient(dbUrl, localMode, basicAuth, true)
```
Comment (gesellix) [inline couchdb-exporter_test.go:356]:
```suggestion
	client := lib.NewCouchdbClient(dbUrl, localOnly, basicAuth, true)
```

File: couchdb-exporter_test.go [hunk 21]
```diff
@@ -334,7 +364,7 @@ func TestCouchdbStatsV2Integration(t *testing.T) {
 
 	scrapeInterval, _ := time.ParseDuration("0s")
 	t.Run("node_up", func(t *testing.T) {
-		e := lib.NewExporter(dbUrl, basicAuth, lib.CollectorConfig{
+		e := lib.NewExporter(dbUrl, localMode, basicAuth, lib.CollectorConfig{
```
Comment (gesellix) [inline couchdb-exporter_test.go:367]:
```suggestion
		e := lib.NewExporter(dbUrl, localOnly, basicAuth, lib.CollectorConfig{
```

File: couchdb-exporter_test.go [hunk 22]
```diff
@@ -359,7 +389,7 @@ func TestCouchdbStatsV2Integration(t *testing.T) {
 	})
 
 	t.Run("_all_dbs", func(t *testing.T) {
-		e := lib.NewExporter(dbUrl, basicAuth, lib.CollectorConfig{
+		e := lib.NewExporter(dbUrl, localMode, basicAuth, lib.CollectorConfig{
```
Comment (gesellix) [inline couchdb-exporter_test.go:392]:
```suggestion
		e := lib.NewExporter(dbUrl, localOnly, basicAuth, lib.CollectorConfig{
```

File: lib/couchdb-client.go [hunk 23]
```diff
@@ -21,6 +21,7 @@ type BasicAuth struct {
 }
 
 type CouchdbClient struct {
+	LocalMode         bool
```
Comment (gesellix) [inline lib/couchdb-client.go:24]:
```suggestion
	LocalOnly         bool
```

File: lib/couchdb-client.go [hunk 24]
```diff
@@ -79,8 +81,12 @@ func (c *CouchdbClient) isCouchDbV1() (bool, error) {
 	return major < 2, nil
 }
 
-func (c *CouchdbClient) GetNodeNames() ([]string, error) {
-	data, err := c.Request("GET", fmt.Sprintf("%s/_membership", c.BaseUri), nil)
+func (c *CouchdbClient) GetNodeNames(localMode bool) ([]string, error) {
```
Comment (gesellix) [inline lib/couchdb-client.go:84]:
```suggestion
func (c *CouchdbClient) GetNodeNames(localOnly bool) ([]string, error) {
```

File: lib/couchdb-client.go [hunk 25]
```diff
@@ -79,8 +81,12 @@ func (c *CouchdbClient) isCouchDbV1() (bool, error) {
...
+	var nodeDiscovery string = "_membership"
+	if localMode{
```
Comment (gesellix) [inline lib/couchdb-client.go:86]:
```suggestion
	if localOnly {
```

File: lib/couchdb-client.go [hunk 26]
```diff
@@ -79,8 +81,12 @@ func (c *CouchdbClient) isCouchDbV1() (bool, error) {
...
+		nodeDiscovery="_node/_local"
+	}
+	data, err := c.Request("GET", fmt.Sprintf("%s/%s", c.BaseUri,nodeDiscovery), nil)
```
Comment (gesellix) [inline lib/couchdb-client.go:89]:
```suggestion
	data, err := c.Request("GET", fmt.Sprintf("%s/%s", c.BaseUri, nodeDiscovery), nil)
```

File: lib/couchdb-client.go [hunk 27]
```diff
@@ -89,14 +95,17 @@ func (c *CouchdbClient) GetNodeNames() ([]string, error) {
 	if err != nil {
 		return nil, err
 	}
+	if localMode{
+		membership.ClusterNodes=append(membership.ClusterNodes,membership.SingleNode)
+	}
```
Comment (gesellix) [inline lib/couchdb-client.go:100]:
```suggestion
	if localOnly {
		membership.ClusterNodes = append(membership.ClusterNodes, membership.SingleNode)
	}
```

File: lib/couchdb-client.go [hunk 28]
```diff
@@ -89,14 +95,17 @@ func (c *CouchdbClient) GetNodeNames() ([]string, error) {
...
 	// for i, name := range membership.ClusterNodes {
 	// 	slog.Infof("node[%d]: %s\n", i, name)
 	// }
 	return membership.ClusterNodes, nil
 }
 
 func (c *CouchdbClient) getNodeBaseUrisByNodeName(baseUri string) (map[string]string, error) {
-	names, err := c.GetNodeNames()
+	names, err := c.GetNodeNames(c.LocalMode)
```
Comment (gesellix) [inline lib/couchdb-client.go:108]:
```suggestion
	names, err := c.GetNodeNames(c.LocalOnly)
```

File: lib/couchdb-client.go [hunk 29]
```diff
@@ -204,7 +213,7 @@ func (c *CouchdbClient) getStats(config CollectorConfig) (Stats, error) {
 		if config.CollectSchedulerJobs {
 			schedulerJobs, err = c.getSchedulerJobs()
 		}
-		activeTasks, err := c.getActiveTasks()
+		activeTasks, err := c.getActiveTasks(c.LocalMode)
```
Comment (gesellix) [inline lib/couchdb-client.go:216]:
```suggestion
		activeTasks, err := c.getActiveTasks(c.LocalOnly)
```

File: lib/couchdb-client.go [hunk 30]
```diff
@@ -514,8 +523,12 @@ func (c *CouchdbClient) getSchedulerJobs() (SchedulerJobsResponse, error) {
 	return schedulerJobs, nil
 }
 
-func (c *CouchdbClient) getActiveTasks() (ActiveTasksResponse, error) {
-	data, err := c.Request("GET", fmt.Sprintf("%s/_active_tasks", c.BaseUri), nil)
+func (c *CouchdbClient) getActiveTasks(localMode bool) (ActiveTasksResponse, error) {
```
Comment (gesellix) [inline lib/couchdb-client.go:526]:
```suggestion
func (c *CouchdbClient) getActiveTasks(localOnly bool) (ActiveTasksResponse, error) {
```

File: lib/couchdb-client.go [hunk 31]
```diff
@@ -514,8 +523,12 @@ func (c *CouchdbClient) getSchedulerJobs() (SchedulerJobsResponse, error) {
...
+	var tasksDiscovery string = "_active_tasks"
+	if localMode {
```
Comment (gesellix) [inline lib/couchdb-client.go:528]:
```suggestion
	if localOnly {
```

File: lib/couchdb-client.go [hunk 32]
```diff
@@ -514,8 +523,12 @@ func (c *CouchdbClient) getSchedulerJobs() (SchedulerJobsResponse, error) {
...
+		tasksDiscovery = "_node/_local/_active_tasks"
+	}
+	data, err := c.Request("GET", fmt.Sprintf("%s/%s", c.BaseUri,tasksDiscovery), nil)
```
Comment (gesellix) [inline lib/couchdb-client.go:531]:
```suggestion
	data, err := c.Request("GET", fmt.Sprintf("%s/%s", c.BaseUri, tasksDiscovery), nil)
```

File: lib/couchdb-client.go [hunk 33]
```diff
@@ -528,7 +541,11 @@ func (c *CouchdbClient) getActiveTasks() (ActiveTasksResponse, error) {
 	for _, activeTask := range activeTasks {
 		// CouchDB 1.x doesn't know anything about nodes.
 		if activeTask.Node == "" {
-			activeTask.Node = "master"
+			if localMode {
+				activeTask.Node = "_local"
+			}else{
+				activeTask.Node = "master"
+			}
```
Comment (gesellix) [inline lib/couchdb-client.go:548]:
```suggestion
			if localOnly {
				activeTask.Node = "_local"
			} else {
				activeTask.Node = "master"
			}
```

File: lib/couchdb-client.go [hunk 34]
```diff
@@ -597,7 +614,7 @@ func (rt *requestCountingRoundTripper) RoundTrip(req *http.Request) (*http.Respo
 	return rt.rt.RoundTrip(req)
 }
 
-func NewCouchdbClient(uri string, basicAuth BasicAuth, insecure bool) *CouchdbClient {
+func NewCouchdbClient(uri string, localMode bool, basicAuth BasicAuth, insecure bool) *CouchdbClient {
```
Comment (gesellix) [inline lib/couchdb-client.go:617]:
```suggestion
func NewCouchdbClient(uri string, localOnly bool, basicAuth BasicAuth, insecure bool) *CouchdbClient {
```

File: lib/couchdb-client.go [hunk 35]
```diff
@@ -611,6 +628,7 @@ func NewCouchdbClient(uri string, basicAuth BasicAuth, insecure bool) *CouchdbCl
 
 	return &CouchdbClient{
 		BaseUri:   uri,
+		LocalMode: localMode,
```
Comment (gesellix) [inline lib/couchdb-client.go:631]:
```suggestion
		LocalOnly: localOnly,
```

File: lib/exporter.go [hunk 36]
```diff
@@ -127,10 +127,10 @@ func (e *Exporter) maybeStartScraping() {
 	}
 }
 
-func NewExporter(uri string, basicAuth BasicAuth, collectorConfig CollectorConfig, insecure bool) *Exporter {
+func NewExporter(uri string, localMode bool, basicAuth BasicAuth, collectorConfig CollectorConfig, insecure bool) *Exporter {
```
Comment (gesellix) [inline lib/exporter.go:130]:
```suggestion
func NewExporter(uri string, localOnly bool, basicAuth BasicAuth, collectorConfig CollectorConfig, insecure bool) *Exporter {
```

File: lib/exporter.go [hunk 37]
```diff
@@ -127,10 +127,10 @@ func (e *Exporter) maybeStartScraping() {
...
 
 	e := &Exporter{
-		client:          NewCouchdbClient(uri, basicAuth, insecure),
+		client:          NewCouchdbClient(uri, localMode, basicAuth, insecure),
```
Comment (gesellix) [inline lib/exporter.go:133]:
```suggestion
		client:          NewCouchdbClient(uri, localOnly, basicAuth, insecure),
```

File: testdata/couchdb-localnode-response-v2.json [hunk 38]
```diff
@@ -0,0 +1,3 @@
+{
+  "name": "node1@127.0.0.1"
+}
```
Comment (gesellix) [inline testdata/couchdb-localnode-response-v2.json:3]:
```suggestion
}

```

**Review by gesellix — COMMENTED** [review]
I left some suggestions to have a more consistent naming. Please check and apply them. I guess you would have to change your current configuration as soon as your changes are release. This is only a heads-up :)

I'll have another look at the implementation later today to give more specific feedback (if any). I'm not yet settled on the idea of a single boolean flag versus a "mode" parameter with multiple possible values.

**Comment by lat0z** [comment]
Hey @gesellix yeah your suggestions makes sense, I've committed all of them, thank you for taking the time to check this proposal

**Review by gesellix — APPROVED** [review]

**Review by gesellix — APPROVED** [review]

**Comment by gesellix** [comment]
It looks like the required checks cannot be started although the pull request has been approved. This is most likely because I've waited too long for the review, so the workflow has been deleted by GitHub.
I checked locally now, and everything looks fine.

**Comment by gesellix** [comment]
@lat0z thanks a lot for your contribution and patience! I've merged your pull request and will create a release later today.

**Comment by gesellix** [comment]
-> https://github.com/gesellix/couchdb-prometheus-exporter/releases/tag/v30.13.0

**Pull request merged**